bootloader/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+  paddr: u32,         // segments physical address
+  filesz: u32,        // size of the segment in file in bytes
+  memsz: u32,         // size of the segment in memory in bytes
+  flags: u32,         // see wikipedia
+  align: u32,         // offset to allign the program on
+}
+
+const SECTSIZE: u32 = 512;
+
+#[panic_handler]
+#[no_mangle]
+pub fn panic(_info: &PanicInfo) -> ! {
+    loop {}
+}
+
+
+// Takes a u32 which represents the location of the elfHeader in memory
+// It then builds a elfHeader on the stack
+// This is unsafe because the caller must ensure "elf_loc" contains an address
+// with a elf_header in it.
+unsafe fn get_elf_header(elf_loc: u32) -> elfHeader {
+    // This should be fairly safe....
+    let elf_ref = mem::transmute::<u32, &mut elfHeader>(elf_loc);
+    elf_ref.clone()
+}
+
+
+// Takes a u32 which represents the location of the progHeader in memory
+// It then builds a elfHeader on the stack
+// This is unsafe because the caller must ensure "prog_loc" contains an address
+// with a elf_header in it.
+unsafe fn get_prog_header(prog_loc: u32) -> progHeader {
+    // This should be fairly safe....
+    let prog_ref = mem::transmute::<u32, &mut progHeader>(prog_loc);
+    prog_ref.clone()
+}
+
+#[no_mangle]
+pub extern "C" fn bootmain() {
+    const ELF_LOC: u32 = 0x10000;
+    readseg(ELF_LOC, 4096, 0);
+    // Copy the elf from the heap to the stack for safe access
+    let elf_header = unsafe{ get_elf_header(ELF_LOC) };
+    if elf_header.magic != ELF_MAGIC {
+        loop {}
+    }
+
+    let mut prog_entry: u32 = 0;
+    while prog_entry < elf_header.phnum as u32 {
+        // For each program header entry, load the data it points to into memory at the appropriate
+        // physical address, we have no virtual memory yet
+        let prog_header_addr = ELF_LOC + elf_header.phoff + prog_entry * mem::size_of::<progHeader>() as u32;
+        let prog_header = unsafe { get_prog_header(prog_header_addr) };
+        let physical_address = prog_header.paddr;
+        readseg(physical_address, prog_header.filesz, prog_header.off);
+
+        // If the entry takes up extra space in memory, fill that extra space with 0s
+        if prog_header.memsz > prog_header.filesz {
+            let zeroes = unsafe {
+                core::slice::from_raw_parts_mut(
+                    (physical_address + prog_header.filesz) as *mut u8,
+                    (prog_header.memsz - prog_header.filesz) as usize
+                )
+            };
+            for i in zeroes {
+                *i = 0;
+            }
+        }
+        prog_entry += 1;
+    }
+
+    // Cast the entry point to a C function pointer and call into the kernel!
+    let entry_ptr = unsafe {
+        let tmp = elf_header.entry as *const ();
+        mem::transmute::<*const (), extern "C" fn()>(tmp)
+    };
+    entry_ptr();
+}
+
+
+/// Wait for disk ready.
+fn waitdisk() {
+
+    let mut not_ready = (get_disk_status() & 0xC0) != 0x40;
```

Comment:
Since `waitdisk()` just waits, do you think using a `loop` might be more self-descriptive here?

```rust
fn waitdisk() {
    loop {
        if (get_disk_status() & 0xC0) != 0x40 { return; // or break }
    }
}
````